libr/anal/fcn.c

```diff
@@ -69,6 +79,126 @@ R_API void r_anal_fcn_update_tinyrange_bbs(RAnalFunction *fcn) {
 	}
 }
 
+// _fcn_tree_{cmp_addr,calc_max_addr,free,probe} are used by interval tree.
+static int _fcn_tree_cmp_addr(const void *a_, const RBNode *b_) {
+	const RAnalFunction *a = (const RAnalFunction *)a_,
+		*b = container_of (b_, const RAnalFunction, rb);
+	ut64 from0 = a->addr, to0 = a->addr + a->_size,
+		from1 = b->addr, to1 = b->addr + b->_size;
+	if (from0 != from1) {
+		return from0 < from1 ? -1 : 1;
+	}
+	if (to0 != to1) {
+		return to0 - 1 < to1 - 1 ? -1 : 1;
+	}
+	return 0;
+}
+
+static void _fcn_tree_calc_max_addr(RBNode *node) {
+	RAnalFunction *fcn = FCN_CONTAINER (node), *fcn1;
+	int i;
+	fcn->rb_max_addr = fcn->addr + fcn->_size - 1;
+	for (i = 0; i < 2; i++) {
+		if (node->child[i]) {
+			fcn1 = container_of (node->child[i], RAnalFunction, rb);
+			if (fcn1->rb_max_addr > fcn->rb_max_addr) {
+				fcn->rb_max_addr = fcn1->rb_max_addr;
+			}
+		}
+	}
+}
+
+static void _fcn_tree_free(RBNode *node) {
+	// TODO after ownership transfered from fcns to fcn_tree
+	// r_anal_fcn_free (FCN_CONTAINER (node));
+}
+
+static RBNode *_fcn_tree_probe(FcnTreeIter *it, RBNode *x_, ut64 from, ut64 to) {
+	RAnalFunction *x = FCN_CONTAINER (x_), *y;
+	RBNode *y_;
+	for (;;) {
+		if ((y_ = x_->child[0]) && (y = FCN_CONTAINER (y_), from <= y->rb_max_addr)) {
+			it->path[it->len++] = x_;
+			x_ = y_;
+			x = y;
+			continue;
+		}
+		if (x->addr <= to - 1) {
+			if (from <= x->addr + x->_size - 1) {
+				return x_;
+			}
+			if ((y_ = x_->child[1])) {
+				x_ = y_;
+				x = FCN_CONTAINER (y_);
+				if (from <= x->rb_max_addr) {
+					continue;
+				}
+			}
+		}
+		return NULL;
+	}
+}
+
+R_API bool r_anal_fcn_tree_delete(RBNode **root, RAnalFunction *data) {
+	return r_rbtree_aug_delete (root, data, _fcn_tree_cmp_addr, _fcn_tree_free, _fcn_tree_calc_max_addr);
+}
+
+R_API void r_anal_fcn_tree_insert(RBNode **root, RAnalFunction *fcn) {
+	r_rbtree_aug_insert (root, fcn, &fcn->rb, _fcn_tree_cmp_addr, _fcn_tree_calc_max_addr);
+}
+
+// Find RAnalFunction whose addr is equal to from
+static RAnalFunction *_fcn_tree_find_addr(RBNode *x_, ut64 from) {
+	while (x_) {
+		RAnalFunction *x = FCN_CONTAINER (x_);
+		if (from < x->addr) {
+			x_ = x_->child[0];
+		} else if (from > x->addr) {
+			x_ = x_->child[1];
```

Comment:
I wuld put the == with return conditional first and the just ternary op to choose the0 or 1 to index the child. Less code and essier to resd